examples/fluids/include/matops.h

```diff
@@ -22,16 +22,21 @@ struct OperatorApplyContext_ {
   Ceed         ceed;
 };
 
-PetscErrorCode MatopApplyContextCreate(DM dm_x, DM dm_y, Ceed ceed, CeedOperator op_apply, CeedVector x_ceed, CeedVector y_ceed, Vec X_loc, Vec Y_loc,
-                                       MatopApplyContext *op_apply_ctx);
-PetscErrorCode MatopApplyContextDestroy(MatopApplyContext op_apply_ctx);
+PetscErrorCode OperatorApplyContextCreate(DM dm_x, DM dm_y, Ceed ceed, CeedOperator op_apply, CeedVector x_ceed, CeedVector y_ceed, Vec X_loc,
+                                          Vec Y_loc, OperatorApplyContext *op_apply_ctx);
+PetscErrorCode OperatorApplyContextDestroy(OperatorApplyContext op_apply_ctx);
 PetscErrorCode MatGetDiag_Ceed(Mat A, Vec D);
-PetscErrorCode ApplyLocal_Ceed(Vec X, Vec Y, MatopApplyContext op_apply_ctx);
 PetscErrorCode MatMult_Ceed(Mat A, Vec X, Vec Y);
 
 PetscErrorCode VecP2C(Vec X_petsc, PetscMemType *mem_type, CeedVector x_ceed);
 PetscErrorCode VecC2P(CeedVector x_ceed, PetscMemType mem_type, Vec X_petsc);
 PetscErrorCode VecReadP2C(Vec X_petsc, PetscMemType *mem_type, CeedVector x_ceed);
 PetscErrorCode VecReadC2P(CeedVector x_ceed, PetscMemType mem_type, Vec X_petsc);
+PetscErrorCode VecCopyP2C(Vec X_petsc, CeedVector x_ceed);
+
+PetscErrorCode ApplyCeedOperatorGlobalToGlobal(Vec X, Vec Y, OperatorApplyContext ctx);
+PetscErrorCode ApplyAddCeedOperatorCeedVectorToLocal(CeedVector x_ceed, Vec Y_loc, OperatorApplyContext ctx);
+PetscErrorCode ApplyCeedOperatorCeedVectorToGlobal(CeedVector x_ceed, Vec Y, OperatorApplyContext ctx);
+PetscErrorCode ApplyAddCeedOperatorGlobalToCeedVector(Vec X, CeedVector y_ceed, OperatorApplyContext ctx);
```

Comment:
Any feedback on these (d00e38d) changes @jedbrown @jeremylt ? Wanted to check in before I continued forward.

Basically took the changes introduced in Ratel here: https://gitlab.com/micromorph/ratel/-/commit/283206d82b406db70993966f9adfff77cab6c05d
and expanded it to allow for CeedVector inputs/outputs and for `CeedOperatorApplyAdd`. Obviously the full tensor-product of global, local, CeedVector, and ApplyAdd would be 18 `ApplyCeedOperator*` functions, but I'd only plan on adding them on an as needed basis.

Reply:
Agreed. Most of the use cases for CeedVector inputs/outputs are for short-lived operators (a vector of ones for creation of a lumped mass matrix) or for data that doesn't live at nodes (spanwise statistics collection). 

I can add a comment onto those functions emphasizing the use of `Vec` over `CeedVector` for manipulation of data when possible.

Reply:
In your mind, would _every_ `CeedVector` have a corresponding `Vec`? Even stuff like `q_data`? Or just operator inputs/outputs?

I agree with the general principal of putting everything in `Vec`s, but am not sure if I agree with/understand that being a hard rule (which I don't think you're advocating for).

Reply:
In my perfect world, an application developer would only need to know about PETSc functions and libCEED QFunctions and nothing else about libCEED. 

We aren't there yet though.

If this is already wrapped in a `OperatorApplyContext` can't you bypass this whole issue by `DMGetLocalVector`ing and setting that local Vec to all 1s?

Reply:
> If this is already wrapped in a `OperatorApplyContext` can't you bypass this whole issue by `DMGetLocalVector`ing and setting that local Vec to all 1s?

In the case of the lumped mass matrix, the issue is one of laziness. Currently, I have to create a `CeedVector`, regardless of which way I go. If I want to use a local `Vec` as the input method, then I need to create both a `CeedVector` and a `Vec`, which is redundant and (pedantically) annoying.

A potential solution to this would be to generate the required `CeedVector` on the fly via, similar to how we use `DMGetLocalVector` to get local `Vec`s on the fly. 

*Sidenote: I previously thought that `CeedVectors` stored information about their data layout, but now realize they only know the size of the data, which makes things much more straight forward (and I think understand your point better too).*

So maybe something like:
```c
PetscErrorCode OperatorApplyContextCreate(DM dm_x, DM dm_y, Ceed ceed, CeedOperator op_apply, CeedVector x_ceed, CeedVector y_ceed, Vec X_loc,
                                          Vec Y_loc, OperatorApplyContext *op_apply_ctx) {
  CeedSize x_size, y_size;

  PetscFunctionBeginUser;
  PetscCall(PetscNew(op_apply_ctx));

  // Copy PETSc Objects
  if (dm_x) PetscCall(PetscObjectReference((PetscObject)dm_x));
  (*op_apply_ctx)->dm_x = dm_x;
  if (dm_y) PetscCall(PetscObjectReference((PetscObject)dm_y));
  (*op_apply_ctx)->dm_y = dm_y;

  if (X_loc) PetscCall(PetscObjectReference((PetscObject)X_loc));
  (*op_apply_ctx)->X_loc = X_loc;
  if (Y_loc) PetscCall(PetscObjectReference((PetscObject)Y_loc));
  (*op_apply_ctx)->Y_loc = Y_loc;

  // Copy libCEED objects
  CeedOperatorGetActiveVectorLengths(op_apply, &x_size, &y_size);
  if (x_ceed) CeedVectorReferenceCopy(x_ceed, &(*op_apply_ctx)->x_ceed);
  else CeedVectorCreate(ceed, x_size, &(*op_apply_ctx)->x_ceed);
  if (y_ceed) CeedVectorReferenceCopy(y_ceed, &(*op_apply_ctx)->y_ceed);
  else CeedVectorCreate(ceed, y_size, &(*op_apply_ctx)->y_ceed);

  CeedOperatorReferenceCopy(op_apply, &(*op_apply_ctx)->op);
  CeedReferenceCopy(ceed, &(*op_apply_ctx)->ceed);

  PetscFunctionReturn(0);
}
```

We could even throw some checks in there for verifying that the size of the local `Vec`s matches the size for the operator. I think that combined with a version of `RatelCeedOperatorCreateLocalVecs` eliminates any desire to have `CeedVector` inputs/outputs.

Reply:
I can probably simplify things a tiny bit in Ratel by making the CeedVectors inside of the ContextCreate function. The length checks I really should add too.

Reply:
For this PR, I've stuck with only the `{Global,Local}To{Global,Local}` functions for right now. I'll do the others in a later PR.

Reply:
> I'll do the others in a later PR.

Opened up #1185